Add a track for Metricbeat data

pcsanwald opened the pull request:
This PR creates a track for metricbeat data, which I'd like to use to benchmark the autohisto aggregation. I've marked this "WIP (work in progress)" because I need to generate a realistic dataset, but, I'd like any feedback on what I've done so far also, as this is the first rally track I've created.

I hope that metricbeat data will be broadly useful, and not just useful for my use case, per the discussion in #19. My idea for generating a dataset is to have one with two metricbeat modules running, that collect beats at different intervals, perhaps an elasticsearch module running hourly, and a system module running on the minute. We'd then run the autohisto aggregation on a field that's only present in elasticsearch (`node_stats` maybe), across a time range and bucket number combination that would result in minute level buckets.

danielmitterdorfer commented on metricbeat/challenges/default.json:
```diff
@@ -0,0 +1,179 @@
+        },
+        {
+          "operation": "index-append",
+          "warmup-time-period": 0,
```
I guess bulk-indexing the document corpus is considered more of a setup task here (i.e. you're not interested in bulk-indexing throughput)? In that case setting the warmup time-period to zero is fine.

danielmitterdorfer commented on metricbeat/challenges/default.json:
```diff
@@ -0,0 +1,179 @@
+          "clients": 1
+        },
+        {
+          "operation": "index-stats",
```
I don't think we need to benchmark index stats here?

danielmitterdorfer commented on metricbeat/challenges/default.json:
```diff
@@ -0,0 +1,179 @@
+          "target-throughput": 100
+        },
+        {
+          "operation": "node-stats",
```
I don't think we need to benchmark node stats here?

danielmitterdorfer commented on metricbeat/challenges/default.json:
```diff
@@ -0,0 +1,179 @@
+      ]
+    },
+    {
+      "name": "append-no-conflicts-index-only",
```
I would not include any of the other challenges *unless* you are interested in those metrics.

danielmitterdorfer commented on metricbeat/challenges/default.json:
```diff
@@ -0,0 +1,179 @@
+          "target-throughput": 100
+        },
+        {
+          "operation": "default",
```
Is it intentional that the aggregation queries are not referenced here?

danielmitterdorfer commented on metricbeat/files.txt:
```diff
@@ -0,0 +1 @@
+documents.json
```
This file is needed for the script that prepares the corpus for offline usage (see https://github.com/elastic/rally-tracks/blob/master/download.sh). I think when you're done this should only contain the compressed corpus file (+ a smaller file that contains the test corpus, see our [docs](https://esrally.readthedocs.io/en/stable/adding_tracks.html#adding-support-for-test-mode)).

danielmitterdorfer replied:
I think this should contain:

```
documents.json.bz2
documents-1k.json.bz2
```

For details please see my comment above.

danielmitterdorfer commented on metricbeat/index.json:
```diff
@@ -0,0 +1,9216 @@
+    "index.search.slowlog.threshold.query.debug" : "0s"
+  },
+  "mappings": {
+    "doc": {
```
Can we use `_doc` here for consistency with the other tracks?

danielmitterdorfer commented on metricbeat/track.json:
```diff
@@ -0,0 +1,32 @@
+    {
+      "name": "metricbeat",
+      "body": "index.json",
+      "types": ["doc"]
```
Can we use `_doc` here for consistency with the other tracks?

danielmitterdorfer commented on metricbeat/track.json:
```diff
@@ -0,0 +1,32 @@
+  "corpora": [
+    {
+      "name": "metricbeat",
+      "base-url": "http://0.0.0.0:8000",
```
This property is optional so you can also remove it while you're iterating.

pcsanwald commented:
@danielmitterdorfer I think I've addressed most of your comments, this is ready for another review I think.

Note that I have gotten an actual slug of metricbeat data from `infra-stats`, so, when we are ready to merge this, I can upload that to S3 and update the config to account for the zipped version, etc.

danielmitterdorfer commented on metricbeat/operations/default.json:
```diff
@@ -0,0 +1,65 @@
+      "ingest-percentage": {{ingest_percentage | default(100)}}
+    },
+    {
+      "name": "default",
```
Are you interested in measuring performance of `match_all`? If not I suggest to remove it from the track.

danielmitterdorfer commented on metricbeat/index.json:
```diff
@@ -0,0 +1,10962 @@
+    "index.search.slowlog.threshold.fetch.debug" : "0s",
+    "index.search.slowlog.threshold.query.debug" : "0s"
+  },
+  "mappings": {
```
Meanwhile types have been removed on master so this will issue deprecation warnings. I have extracted the Metricbeat mapping from 7.0.0-beta1 in https://gist.github.com/danielmitterdorfer/72686146100dea7b775f0efdc80bd2f1. Can you please use that mapping instead so we avoid the deprecation warnings?

danielmitterdorfer commented on metricbeat/track.json:
```diff
@@ -0,0 +1,31 @@
+    {
+      "name": "metricbeat",
+      "body": "index.json",
+      "types": ["_doc"]
```
As types have been removed on master you can just remove this property (the most recent version of Rally will allow you to do this).

pcsanwald commented:
certainly no urgency on this, but I think I've addressed review comments and this should be ready for another review.

pcsanwald commented:
I've pinged daniel and we are uploading the track data

danielmitterdorfer commented on metricbeat/track.json:
```diff
@@ -0,0 +1,30 @@
+  ],
+  "corpora": [
+    {
+      "name": "metricbeat",
```
can you please add the following property:

```json
"base-url": "http://benchmarks.elasticsearch.org.s3.amazonaws.com/corpora/metricbeat"
```

pcsanwald replied:
I pushed b9326

danielmitterdorfer commented on metricbeat/track.json:
```diff
@@ -0,0 +1,30 @@
+      "name": "metricbeat",
+      "documents": [
+        {
+          "source-file": "documents.json",
```
This should be `documents.json.bz2` so it is using the compressed version.

danielmitterdorfer commented:
The track data are now uploaded to S3 and I tested it locally. I left two more comments (see above). Can you please address them @pcsanwald? Then I think we're good to merge.

pcsanwald commented:
@danielmitterdorfer thanks for the review! I believe I have addressed both your comments.
